src/build_backend.py

```diff
@@ -0,0 +1,119 @@
+import argparse
+import hashlib
+import os
+import shutil
+import subprocess
+import tarfile
+import zipfile
+
+from typing import Dict, Iterable, Optional
+
+from cockpit import __version__
+
+PACKAGE = f'cockpit-{__version__}'
+TAG = 'py3-none-any'
+
+
+def find_sources(srcpkg: bool) -> Iterable[str]:
+    try:
+        subprocess.check_call(['modules/checkout'], stdout=2)
+    except FileNotFoundError:
+        pass
+
+    if srcpkg:
+        yield from {
+            'pyproject.toml',
+            'src/build_backend.py',
+        }
+
+    for path, _dirs, files in os.walk('src', followlinks=True):
+        if '__init__.py' in files:
+            yield from [os.path.join(path, file) for file in files]
+
+
+def copy_sources(distdir: str) -> None:
+    for source in find_sources(srcpkg=True):
+        destination = os.path.join(distdir, source)
+        os.makedirs(os.path.dirname(destination), exist_ok=True)
+        shutil.copy(source, destination)
+
+
+def build_sdist(sdist_directory: str,
+                config_settings: Optional[Dict[str, object]] = None) -> str:
+    del config_settings
+    sdist_filename = f'{PACKAGE}.tar.gz'
+    with tarfile.open(f'{sdist_directory}/{sdist_filename}', 'w:gz', dereference=True) as sdist:
+        for filename in find_sources(srcpkg=True):
+            sdist.add(filename, arcname=f'{PACKAGE}/{filename}', )
+    return sdist_filename
+
+
+def build_wheel(wheel_directory: str,
+                config_settings: Optional[Dict[str, object]] = None,
+                metadata_directory: Optional[str] = None) -> str:
+    del config_settings, metadata_directory
+    wheel_filename = f'{PACKAGE}-{TAG}.whl'
+    record_file = f'{PACKAGE}.dist-info/RECORD'
+    distinfo = {
+        'WHEEL': [
+            'Wheel-Version: 1.0',
+            'Generator: cockpit build_backend',
+            'Root-Is-Purelib: true',
+            f'Tag: {TAG}',
+        ],
+        'METADATA': [
+            'Metadata-Version: 2.1',
+            'Name: cockpit',
+            f'Version: {__version__}',
+        ],
+        'entry_points.txt': [
+            '[console_scripts]',
+            'cockpit-bridge = cockpit.bridge:main',
+            'cockpit-askpass = cockpit._vendor.ferny.interaction_client:main',
+        ],
+    }
+
+    with zipfile.ZipFile(f'{wheel_directory}/{wheel_filename}', 'w') as wheel:
+        for filename in find_sources(srcpkg=False):
+            wheel.write(filename, arcname=os.path.relpath(filename, start='src'))
+
+        for filename, contents in distinfo.items():
+            wheel.writestr(f'{PACKAGE}.dist-info/{filename}', ''.join(f'{line}\n' for line in contents))
```

Comment:
How is that different from ``\n`.join(contents)`?

Reply:
It differs in the trailing newline.  It's equivalent to something like

```python
  '\n'.join(contents) + '\n'
```

but to my eye is slightly cleaner.